src/improc.c

```diff
@@ -2512,3 +2512,175 @@ DoubleImage ifft2DDouble(ComplexImage image) {
+* @param out The 1D int array that the result of the operation will be stored in
+* @param n The size of both of the arrays (in nr of elements)
+* @param w The size of the window
+* @param ord If `ord` == 1, the sliding minimum is computed. Otherwise a sliding maximum is computed
+* @param offset Determines the gap in position in the input/output arrays between two successive elements
+* @param start Determines the position of the first element in the input/output arrays
+* @param workspace An array of size w (in nr of elements) to use as backing storage for our internal datastructure.
+*/
+static void slidingWindowOrd(int *img, int *out, int n, int w, int ord, int offset, int start, int *workspace) {
+  // our quack datastructure stores indices of numers in our original array
+  Quack quack = createNewQuackWithMemory(w, workspace);
+
+  for (int i = 0; i < n; i++) {
+    // remove any elements from the front (oldest side) of the quack that are now out of the "reach" of our window
+    while (!quackIsEmpty(&quack) && quackPeekFront(&quack) <= i - w) {
+      quackPopFront(&quack);
+    }
+
+    // we are now observing the current value, img[i]. we pop values from the back (most recent side) of the quack
+    // until we encounter a value that is larger (and makes the current value insignificant). our loop invariant is
+    // that we keep the largest value currently in our window at the front of the quack.
+    // NOTE: the comparison to ord is a complicated way of switching between < and >
+    while (!quackIsEmpty(&quack) && (img[quackPeekBack(&quack) * offset + start] <= img[i * offset + start]) == ord) {
+      quackPopBack(&quack);
+    }
+
+    quackPushBack(&quack, i);
+
+    out[i * offset + start] = img[quackPeekFront(&quack) * offset + start];
+  }
+}
+
+/**
+* Perform either a dilation or an erosion (depending on flag) on the input image. The structuring element that will be
+* used will be a rectangle of width `kw` and height `kh`.
+*
+* @param image The image that the dilation or erosion will be applied on.
+* @param kw The width of the rectangular structuring element (kernel)
+* @param kh The heigth of the rectangular structuring element (kernel)
+* @param flag If `flag` == 1, a dilation will be performed. Otherwise an erosion will be performed.
+*/
+IntImage dilateErodeIntImageRect(IntImage image, int kw, int kh, int flag) {
+  ImageDomain domain = getIntImageDomain(image);
+  IntImage result = allocateIntImageGridDomain(domain, image.minRange, image.maxRange);
+  int width, height;
+  getWidthHeight(domain, &width, &height);
+
+  // we allocate memory here to avoid having to repeat allocations for every row/col
+  int largestDim = maxOp(kw, kh);
+  int *memory = malloc(largestDim * sizeof(*memory));
+
+  // first we run the min/max operation on the image row-wise, saving the sliding window min/max in the result image
+  for (int row = 0; row < height; row++) {
```

Comment:
Had some issues with the domain so I opted for reading width and height directly.. This is "dangerous" accessing but I think it's fine. Maybe we should use the getIntImagePixelI functions instead?

Reply:
In general, I would do most things using width/height instead of the domain, unless you require unique indexing. Technically using `getIntImagePixelI` would be safer, but if it works like this, it's fine I suppose.